tests/target/issue-binop-comments/binop-comments-front.rs

```diff
@@ -0,0 +1,298 @@
+        && context /* condition 3 expression */
+        && !res.ends('+')
+    {
+        a + b
+    }
+}
+// Test 12.2 - Pre-comments break to few lines - "&&" and "||"
+fn main() {
+    if context() /* condition 1 expression */ && bounds.len() == 1 /* condition 2 expression */
+        || context /* condition 3 expression */ && !res.ends('+')
+    {
+        a + b
+    }
+}
+
+// Test 13.1 - Post-comments break to few lines - only "&&"
+fn main() {
+    if context()
+        && /* condition 1 expression */ bounds.len() == 1
+        && /* condition 2 expression */ context
+        && /* condition 3 expression */ !res.ends('+')
+    /* closing comment */
+    {
+        a + b
+    }
+}
+// Test 13.2 - Pre and Post-comments break to few lines - only "&&"
+fn main() {
+    if context() /* pre comment 1 */
+        && /* post comment 1 */ bounds.len() == 1 /* pre comment 2 */
+        && /* post comment 2 */ context /* pre comment 3 */
+        && /* post comment 3 */ !res.ends('+')
+    /* closing comment */
+    {
+        a + b
+    }
+}
+
+// Test 14.1 - Each "&&" starts new line
+fn main() {
+    if context.inside_macro()
+        && bounds.len() == 1
+        && context.snippet(self.span).ends_with('+') & !res.ends_with('+')
+    {
+        a + b
+    }
+}
+// Test 14.2 - "||" starts new line
+fn main() {
+    if context.inside_macro() && bounds.len() == 1
+        || context.snippet(self.span).ends_with('+') & !res.ends_with('+')
+    {
+        a + b
+    }
+}
+
+// Test 15.1 - Each "&&" starts new line - with comments
+fn main() {
+    if context.inside_macro() /* cond 1 */
+        && bounds.len() == 1 /* cond 2 */
+        && context.snippet(self.span).ends_with('+') /* cond 3 */
+        && !res.ends_with('+')
+    {
+        a + b
+    }
+}
+// Test 15.2 - "||" starts new line - with comments
+fn main() {
+    if context.inside_macro() /* cond 1 */ && bounds.len() == 1 /* cond 2 */
+        || context.snippet(self.span).ends_with('+') /* cond 3 */ && !res.ends_with('+')
+    {
+        a + b
+    }
+}
+
+// Test 16 - Each "." and some "&&"/"||" starts new line
+fn main() {
+    if a {
+        if b {
+            if c {
+                if d {
+                    if e {
+                        if item_kind
+                            .is_same_item_kind(&***ppi, self.file_mod_map)
+                            .context
+                            .inside_macro()
+                            && bounds.len() == 1
+                            || context.snippet(self.span).ends_with('+') && !res.ends_with('+')
+                        {
+                            a + b
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Test 21.1
+fn main() {
+    let x = 2 /* v */
+              /* w */
+        * /* x */ 4;
+}
+
+// Test 21.2
+fn main() {
+    let x = 2 /* v */
+              /* w */
+        * /* x */
+          /* y */ 4;
+}
```

Comment:
This one also feels off to me (requirements from the style guide for binops for reference: https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/expressions.md#binary-operations)

I'm not sure if it should be something like 
```rust
fn main() {
    let x = 2 /* v */
        /* w */
        * /* x */
        /* y */ 4;
}
```

or even 

```rust
fn main() {
    let x = 2 /* v */ /* w */
        * /* x */ /* y */ 4;
}
```

but I don't think we should force the operand/rhs to a separate newline with increased indent to support comment alignment

i feel like the correct/expected formatting for some of these snippets is getting a little fuzzy, so cc @topecongiro for some clarity and guidance

Reply:
(cc:  @topecongiro)
The above formatting issues are actually **not** because the pre/post binop comments which this PR is about, but because of outer comments in the expression.  However, they may need to be fixed in this PR, as before the original source of the expression was copied and now the output is the formatted expression, which may be worse if these issues are not fixed.

Before I submit further enhancements, I tried to come up with better examples, so we can decide about the desired output.  Initially the examples are for the _Front_ case only.  In the following examples, I assumed that concatenated closed-comments like `/* v */  /* w */` (maybe with new lines between the two comments) **will no happen in practice**, so they can be **ignored** in the tests cases.  Is that **acceptable?**

Following is what I assume is a **desired output for pre/post binop comments only** (except for a following comment/expression just to  make sure its formatting continues to be o.k.):
```rust
// Desired output (Front)
fn main() {
    let y = 2 /* 1 This is the first bin op argument */
              /* 2 which will be multiplied by the second argument */
        * /* 3 This is the multiplying operation */
          /* 4 which multiplies both arguments */ 4;
    /* 7 And this is a comment for the next operation */
    let i = 7;
}
```
If it is agreed that this formatting is o.k., then the current PR is o.k.

The following is what I assume is the **desired output when outer comments are added**:
```rust
// Desired output (Front)
fn main () {
    let y = 2 /* 1 This is the first bin op argument */
              /* 2 which will be multiplied by the second argument */
         * /* 3 This is the multiplying operation */
           /* 4 which multiplies both arguments */
         4; /* 5 This is the second argument */
            /* 6 which is multiplied by the first argument */
    /* 7 And this is a comment for the next operation */
    let i = 7;
}
```
Currently, there are at least three issues with this example.  After we agree about the desired output, **should I try to fix them i this PR?**

The first issue is that "`4`" doesn't start in a new line, **which is according the current formatting rules**, so it may be that the desired output should be different.  The second issues is that comment 6 is not properly indented and is formatted as comment 7:
```rust
        * /* 3 This is the multiplying operation */
          /* 4 which multiplies both arguments */ 4; /* 5 This is the second argument */
    /* 6 which is multiplied by the first argument */
    /* 7 And this is a comment for the next operation */
    let i = 7;
```
I am not sure if the second issue can be fixed easily, as the original formatting should be taken into account which may be quite complex to do.

The second issue is that when comments 5 and 6 are put before the "`;`", there is no distinction in the formatting between comments 5 and 6 which are for the binop and between comment 7.  
```rsut
        * /* 3 This is the multiplying operation */
          /* 4 which multiplies both arguments */ 4
    /* 5 This is the second argument */
    /* 6 which is multiplied by the first argument */;
    /* 7 And this is a comment for the next operation */
    let i = 7;
```
I assume this issue can be fixed.

Reply:
This is definitely something that would need to be addressed in order for this PR to be merged. Currently rustfmt will leave binary expressions with comments between the operator and operand(s) alone, and allow the programmer to adjust the formatting as needed. However, this patch would remove that, and forcibly rewrite code, so the code that's emitted needs to be correct. 

The biggest thing that jumps out to me with the proposed changes are the forced vertical/visual alignment of the block comments, which is not compatible with the Rust Style Guide, and accordingly, nor rustfmt's defaults. 

For example, the latest version in this PR currently changes:
```rust
fn main() {
    let y = 2 /* 1 This is the first bin op argument */
  /* 2 which will be multiplied by the second argument */
        * /* 3 This is the multiplying operation */ 
                             /* 4 which multiplies both arguments */ 4;
}
```

to
```rust
fn main() {
    let y = 2 /* 1 This is the first bin op argument */
              /* 2 which will be multiplied by the second argument */
        * /* 3 This is the multiplying operation */
          /* 4 which multiplies both arguments */ 4;
}
```

This is not correct

Reply:
> This is definitely something that would need to be addressed in order for this PR to be merged. Currently rustfmt will leave binary expressions with comments between the operator and operand(s) alone, and allow the programmer to adjust the formatting as needed. However, this patch would remove that, and forcibly rewrite code, so the code that's emitted needs to be correct.

Seems that I completely misunderstood the required formatting of multi-line comments.  I understand from the above that multi-line comments should keep their relative indentation.  This will actually solve the indentation issues in the previous message.  I will try to change the PR so it will work according to the above guidelines.  Just to make sure, I understand that for example:
```rust
let y = 2 /* comment 1 */
/* comment 2 */
                       /* comment 3 */
* /* comment 4 */
/* comment 5 */
4;
```
Should be formatted to:
```rust
let y = 2 /* comment 1 */
/* comment 2 */
                       /* comment 3 */
    * /* comment 4 */
    /* comment 5 */
    4;
```

Reply:
Note that there's a difference between having multiple, single-line block comments (as is the case in the snippets discussed here) vs. a single, multi-line block comment.

```rust
/* c1 */     /* c2 */
/* c3 */
```
vs.
```rust
/*
 * c1
 * c2
 * c3
 */
```

Yes the alignment of the multiline comment is important, but we don't need to visually/vertically align separate comments with each other.

> Should be formatted to:
> 
> ```rust
> let y = 2 /* comment 1 */
> /* comment 2 */
>                        /* comment 3 */
>     * /* comment 4 */
>     /* comment 5 */
>     4;
> ```

No, not that either, as rustfmt shouldn't be maintaining incorrect indentation either;  the comments should be properly indented too.

I'd think we'd expect something like this for your example which maintains the author's original comment line positioning, but corrects the indentation for comment 2 and 3
```rust
let y = 2 /* comment 1 */
    /* comment 2 */
    /* comment 3 */
    * /* comment 4 */
    /* comment 5 */ 4;
```

It's a situation like this one that I feel like there may be some ambiguity (all with default behavior/config):
```rust
fn main() {
    let y = 2 /* comment 1 */ /* loooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnn */ * /* another lonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggg 4 */  /* comment 5 */ 4;
}
```
Arguably rustfmt should just fail to format and differ to the programmer to update/correct, because the comments between the lhs operand and operator would exceed max_width, and the comments between the operator and rhs operand will cause that line to also exceed max width. However, I suppose there's a question of whether rustfmt should attempt to split those lines to fit within the limits, and if so, that also begs the question of what to do with the rhs operand given that it's no longer on the same line as the operator (should it go on its own newline as well, should it get another block indent, etc.):

i.e. 
```rust
fn main() {
    let y = 2 /* comment 1 */
        /* loooooooooooooooooooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnn */
        * /* another lonnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggg 4 */
        /* comment 5 */ 4;
}
```

This isn't called out explicitly in the style guide which is unsurprising as these seem like rather edge-y cases.

Reply:
I didn't distinguish between multiline comments and block of separate comments ... I make changes according the the above guidelines, which hopefully will produce output that is closer to the desired result.